operator/endpointslicegc/gc.go

```
@@ -0,0 +1,137 @@
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cilium_api_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// params contains all the dependencies for the endpoint-slice-gc.
+// They will be provided through dependency injection.
+type params struct {
+	cell.In
+
+	Logger    *slog.Logger
+	Lifecycle cell.Lifecycle
+	JobGroup  job.Group
+
+	Clientset            k8sClient.Clientset
+	CiliumEndpointSlices resource.Resource[*cilium_api_v2a1.CiliumEndpointSlice]
+
+	SharedCfg SharedConfig
+}
+
+// GC represents the Cilium Endpoint Slice one-off GC.
+type GC struct {
+	logger *slog.Logger
+
+	clientset            k8sClient.Clientset
+	ciliumEndpointSlices resource.Resource[*cilium_api_v2a1.CiliumEndpointSlice]
+}
+
+func registerGC(p params) {
+	if !p.Clientset.IsEnabled() {
+		return
+	}
+
+	if p.SharedCfg.EnableCiliumEndpointSlice {
+		return
+	}
+
+	gc := &GC{
+		logger:               p.Logger,
+		clientset:            p.Clientset,
+		ciliumEndpointSlices: p.CiliumEndpointSlices,
+	}
+
+	p.JobGroup.Add(
+		job.OneShot(
+			"to-k8s-ciliumendpointslice-gc",
+			gc.doGC,
+			job.WithRetry(3, &job.ExponentialBackoff{
+				Min: 30 * time.Second,
+				Max: 120 * time.Second,
+			}),
+		),
+	)
+}
+
+func (g *GC) checkForCiliumEndpointSliceCRD(ctx cell.HookContext) bool {
+	_, err := g.clientset.ApiextensionsV1().CustomResourceDefinitions().Get(
+		ctx, cilium_api_v2a1.CESName, metav1.GetOptions{ResourceVersion: "0"},
+	)
+	if err == nil {
+		return true
+	} else if k8serrors.IsNotFound(err) {
+		g.logger.InfoContext(ctx, "CiliumEndpointSlice CRD cannot be found, skipping garbage collection", logfields.Error, err)
+	} else {
+		g.logger.ErrorContext(ctx, "Unable to determine if CiliumEndpointSlice CRD is installed, cannot start garbage collector",
```

This should probably return an error, so the job is retried. I assume this will fail only because the apiserver is taking a nap.

---

Fair call, thanks!